File: frontend/src/services/encryptionService.js
/**
 * Frontend Encryption Service
 * Handles decryption of encrypted API responses
 * Uses Fernet-compatible decryption (Python cryptography library compatible)
 */

class EncryptionService {
  constructor() {
    // Get encryption key from environment variable
    this.encryptionKey = import.meta.env.VITE_ENCRYPTION_KEY || '';
    this.encryptionEnabled = import.meta.env.VITE_ENABLE_ENCRYPTION !== 'false';
    
    if (this.encryptionEnabled && !this.encryptionKey) {
      if (import.meta.env.DEV) { console.warn('⚠️ Encryption is enabled but no VITE_ENCRYPTION_KEY found in environment'); }
    }
    
    if (this.encryptionEnabled) {
      if (import.meta.env.DEV) { console.log('✅ Frontend encryption service initialized'); }
    } else {
      if (import.meta.env.DEV) { console.log('ℹ️ Frontend encryption is disabled'); }
    }
  }

  /**
   * Check if response is encrypted
   * @param {Object} response - API response
   * @returns {boolean}
   */
  isEncrypted(response) {
    return response && typeof response === 'object' && response.encrypted === true;
  }

  /**
   * Base64 URL decode (Fernet uses URL-safe base64)
   * @param {string} str - Base64 URL encoded string
   * @returns {Uint8Array}
   */
  base64UrlDecode(str) {
    // Convert URL-safe base64 to standard base64
    str = str.replace(/-/g, '+').replace(/_/g, '/');
    
    // Add padding if needed
    while (str.length % 4) {
      str += '=';
    }
    
    // Decode base64
    const binaryString = atob(str);
    const bytes = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
      bytes[i] = binaryString.charCodeAt(i);
    }
    return bytes;
  }

  /**
   * Decrypt Fernet token
   * @param {string} token - Fernet token (base64 encoded)
   * @param {Uint8Array} key - Encryption key
   * @returns {Promise<string>} - Decrypted plaintext
   */
  async decryptFernet(token, key) {
    try {
      // Decode the token
      const tokenBytes = this.base64UrlDecode(token);
      
      // Fernet token structure:
      // Version (1 byte) | Timestamp (8 bytes) | IV (16 bytes) | Ciphertext (variable) | HMAC (32 bytes)
      
      if (tokenBytes.length < 57) {
        throw new Error('Invalid Fernet token: too short');
      }
      
      const version = tokenBytes[0];
      if (version !== 0x80) {
        throw new Error(`Invalid Fernet version: ${version}`);
      }
      
      // Extract components
      const iv = tokenBytes.slice(9, 25);
      const ciphertext = tokenBytes.slice(25, tokenBytes.length - 32);
      
      // Derive encryption key from Fernet key (first 16 bytes for encryption)
      const encryptionKey = key.slice(0, 16);
      
      // Import key for AES-128-CBC
      const cryptoKey = await crypto.subtle.importKey(
        'raw',
        encryptionKey,
        { name: 'AES-CBC' },
        false,
        ['decrypt']
      );
      
      // Decrypt
      const decrypted = await crypto.subtle.decrypt(
        { name: 'AES-CBC', iv: iv },
        cryptoKey,
        ciphertext
      );
      
      // Convert to string
      const decoder = new TextDecoder();
      return decoder.decode(decrypted);
      
    } catch (error) {
      if (import.meta.env.DEV) { console.error('❌ Fernet decryption error:', error); }
      throw error;
    }
  }

  /**
   * Decrypt API response
   * @param {string} encryptedData - Base64 encoded Fernet token
   * @returns {Promise<any>} - Decrypted data
   */
  async decrypt(encryptedData) {
    if (!this.encryptionEnabled) {
      throw new Error('Encryption is not enabled');
    }

    if (!this.encryptionKey) {
      throw new Error('Encryption key not configured');
    }

    try {
      // Decode the Fernet key
      const keyBytes = this.base64UrlDecode(this.encryptionKey);
      
      // Decrypt the Fernet token
      const decryptedText = await this.decryptFernet(encryptedData, keyBytes);
      
      // Parse JSON
      const data = JSON.parse(decryptedText);
      
      console.debug('✅ Successfully decrypted response data');
      return data;
      
    } catch (error) {
      if (import.meta.env.DEV) { console.error('❌ Decryption failed:', error); }
      throw new Error(`Failed to decrypt response: ${error.message}`);
    }
  }

  /**
   * Process API response - decrypt if encrypted, return as-is if not
   * @param {Object} response - API response
   * @returns {Promise<any>} - Processed data
   */
  async processResponse(response) {
    // Check if response is encrypted
    if (this.isEncrypted(response)) {
      console.debug('🔐 Response is encrypted, decrypting...');
      return await this.decrypt(response.data);
    }
    
    // If encryption is disabled or response is not encrypted
    if (response && typeof response === 'object' && 'data' in response && response.encrypted === false) {
      console.debug('ℹ️ Response is not encrypted (encryption disabled on server)');
      return response.data;
    }
    
    // Return as-is if not in encrypted format
    console.debug('ℹ️ Response is not in encrypted format, returning as-is');
    return response;
  }

  /**
   * Check if encryption is enabled
   * @returns {boolean}
   */
  isEnabled() {
    return this.encryptionEnabled && !!this.encryptionKey;
  }
}

// Create singleton instance
const encryptionService = new EncryptionService();

export default encryptionService;

File: backend/app/services/security/encryption_service.py
"""
Encryption Service for API Response Encryption
Encrypts all API responses so data is not visible in browser Network tab
Uses AES-256 encryption with Fernet
"""
from cryptography.fernet import Fernet
from typing import Any, Dict
import base64
import os
import json
import logging
from datetime import datetime
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

logger = logging.getLogger(__name__)


class EncryptionService:
    """
    Handles encryption and decryption of API responses
    Uses AES-256 encryption with Fernet
    """
    
    def __init__(self):
        """Initialize encryption service with key"""
        self.encryption_enabled = os.getenv('ENABLE_RESPONSE_ENCRYPTION', 'true').lower() == 'true'
        
        if self.encryption_enabled:
            # Get encryption key from environment or generate one
            encryption_key = os.getenv('ENCRYPTION_KEY')
            
            if not encryption_key:
                # Generate a key if not provided
                logger.warning("⚠️ No ENCRYPTION_KEY found in environment. Generating a new key.")
                logger.warning("⚠️ Add this to your .env file: ENCRYPTION_KEY=" + Fernet.generate_key().decode())
                encryption_key = Fernet.generate_key().decode()
            
            # Convert string key to bytes if needed
            if isinstance(encryption_key, str):
                encryption_key = encryption_key.encode()
            
            # Initialize Fernet cipher
            try:
                self.cipher = Fernet(encryption_key)
                logger.info("✅ Encryption service initialized successfully")
            except Exception as e:
                logger.error(f"❌ Failed to initialize encryption: {e}")
                logger.warning("⚠️ Disabling encryption due to initialization error")
                self.encryption_enabled = False
                self.cipher = None
        else:
            logger.info("ℹ️ Response encryption is disabled")
            self.cipher = None
    
    def encrypt_response(self, data: Any) -> Dict[str, str]:
        """
        Encrypt API response data
        
        Args:
            data: Any JSON-serializable data (dict, list, string, etc.)
        
        Returns:
            Dict with encrypted data and metadata
        """
        if not self.encryption_enabled or not self.cipher:
            # Return data as-is if encryption is disabled
            return {
                "encrypted": False,
                "data": data
            }
        
        try:
            # Convert data to JSON string
            json_data = json.dumps(data, default=str)
            
            # Encrypt the JSON string
            encrypted_bytes = self.cipher.encrypt(json_data.encode('utf-8'))
            
            # Convert to base64 for safe transmission
            encrypted_base64 = base64.b64encode(encrypted_bytes).decode('utf-8')
            
            logger.debug(f"✅ Encrypted response data ({len(json_data)} bytes -> {len(encrypted_base64)} bytes)")
            
            return {
                "encrypted": True,
                "data": encrypted_base64,
                "timestamp": datetime.now().isoformat()
            }
            
        except Exception as e:
            logger.error(f"❌ Encryption failed: {e}")
            # Return unencrypted data if encryption fails
            return {
                "encrypted": False,
                "data": data,
                "error": "Encryption failed"
            }
    
    def decrypt_response(self, encrypted_data: str) -> Any:
        """
        Decrypt API response data
        
        Args:
            encrypted_data: Base64-encoded encrypted string
        
        Returns:
            Decrypted data (original format)
        """
        if not self.encryption_enabled or not self.cipher:
            raise ValueError("Encryption is not enabled")
        
        try:
            # Decode from base64
            encrypted_bytes = base64.b64decode(encrypted_data.encode('utf-8'))
            
            # Decrypt
            decrypted_bytes = self.cipher.decrypt(encrypted_bytes)
            
            # Convert back to JSON
            json_data = decrypted_bytes.decode('utf-8')
            data = json.loads(json_data)
            
            logger.debug(f"✅ Decrypted response data")
            
            return data
            
        except Exception as e:
            logger.error(f"❌ Decryption failed: {e}")
            raise ValueError(f"Failed to decrypt data: {str(e)}")
    
    def is_enabled(self) -> bool:
        """Check if encryption is enabled"""
        return self.encryption_enabled and self.cipher is not None


# Global encryption service instance
encryption_service = EncryptionService()


def encrypt_api_response(data: Any) -> Dict[str, Any]:
    """
    Helper function to encrypt API response
    Use this in route handlers to encrypt responses
    """
    return encryption_service.encrypt_response(data)


def decrypt_api_response(encrypted_data: str) -> Any:
    """
    Helper function to decrypt API response
    Use this in frontend to decrypt responses
    """
    return encryption_service.decrypt_response(encrypted_data)
